### PR TITLE
add support to purge apim

### LINKED ISSUE #1185: Add option to purge APIM like we do for Key Vault
**Output from `azd version`**
azd version 0.4.0-beta.1 (commit 2bf7a529914d86d8e9f5d7b6f95a4f20a2351e01)

**Output from `az version`** 
2.42.0

**Describe the bug**
Context: Azure API Management (APIM) may be optionally used to mediate the calls between the web frontend and the backend API.
When we do `azd up` the APIM resource is deployed.
When we do `azd down` the APIM resource is deleted but keeps in a "soft delete" state. So, if we run `azd up` again it will fail because 
 the resource already exists.
We need to implement the purge mechanism by calling the REST API (https://learn.microsoft.com/en-us/azure/api-management/soft-delete) or az CLI (https://learn.microsoft.com/en-us/cli/azure/apim/deletedservice?view=azure-cli-latest#az-apim-deletedservice-purge)

**To Reproduce**
`azd init -t Azure-Samples/todo-csharp-cosmos-sql -b pr/1130`
`set USE_APIM=true`
`azd up`
`azd down`
`azd up` fails with the message:
"ServiceAlreadyExistsInSoftDeletedState: Api service apim-tqy5pg354bzug was soft-deleted. In order to create the new service with the same name, you have to either undelete the service or purge it. See https://aka.ms/apimsoftdelete.
%!s(<nil>)"

**Expected behavior**
After doing `azd down` if an APIM resource was deleted I expect the CLI to ask if I want to purge the resource instance. The same way that happens with Key Vault:
"? Would you like to permanently delete these Key Vaults instead, allowing their names to be reused? (y/N)"

And if I use the --purge switch I will not be prompted with a decision and the purge will be automatically.
After that I can do `azd up` again with success.

**Environment**
Information on your environment:
OS: Windows desktop
Template: https://github.com/Azure-Samples/todo-csharp-cosmos-sql
Branch: https://github.com/Azure/azure-dev/pull/1130

**Additional context**
Maybe we should have an automatic way to understand if a resource type has the soft delete state to make this behavior automatic for all the resources.

### PR DESCRIPTION
**Pull request by hemarina**
fix #1185

### CODE REVIEW

File: cli/azd/pkg/infra/provisioning/bicep/bicep_provider.go [hunk 1]
```diff
@@ -585,6 +599,28 @@ func (p *BicepProvider) getAppConfigsToPurge(
 	return configs, nil
 }
 
+func (p *BicepProvider) getAPIManagementsToPurge(
+	ctx context.Context,
+	groupedResources map[string][]azcli.AzCliResource,
+) ([]*azcli.AzCliAPIM, error) {
+	apims := []*azcli.AzCliAPIM{}
+
+	for resourceGroup, groupResources := range groupedResources {
```
Comment (ellismg) [inline cli/azd/pkg/infra/provisioning/bicep/bicep_provider.go:608]: Might be worthwhile to add a comment here to note that soft-delete happens for all API Management resources, and that's why we don't have an additional level of filtering like we do on some of our other functions.

File: cli/azd/pkg/infra/provisioning/bicep/bicep_provider.go [hunk 2]
```diff
@@ -282,6 +282,13 @@ func (p *BicepProvider) Destroy(
 				return
 			}
 
+			asyncContext.SetProgress(&DestroyProgress{Message: "Getting API Managements to purge", Timestamp: time.Now()})
```
Comment (ellismg) [inline cli/azd/pkg/infra/provisioning/bicep/bicep_provider.go:285]: Maybe "Getting API Management Services to purge"?   I think Azure calls these an instance of the resource a "API Management Service", and it feels like it reads a little better.

File: cli/azd/pkg/infra/provisioning/bicep/bicep_provider.go [hunk 3]
```diff
@@ -632,6 +668,42 @@ func (p *BicepProvider) purgeAppConfigs(
...
+	ctx context.Context,
+	asyncContext *async.InteractiveTaskContextWithProgress[*DestroyResult, *DestroyProgress],
+	apims []*azcli.AzCliAPIM,
+	options DestroyOptions,
+) error {
+	for _, apim := range apims {
+		progressReport := DestroyProgress{
+			Timestamp: time.Now(),
+			Message: fmt.Sprintf(
+				"%s api management %s",
```
Comment (ellismg) [inline cli/azd/pkg/infra/provisioning/bicep/bicep_provider.go:681]: maybe "api management service" here as well.

**Review by ellismg — APPROVED** [review]
Some small feedback on some prose in messages but otherwise looks great.  Thanks for fixing this!

**Review by wbreza — APPROVED** [review]
This looks good enough for now, but seems like we'll be coming back here multiple times to add support for each new resource that we want to support as part of the purge flow.

It might be a good idea to create a new issues to refactor this so it is easier to add new support in the future where we can introduce a concept of a `PurgeableResource` or similar where we can register the APIs required to collect, validate and purge the resource but all wraps up into a easy to consume interface.

File: cli/azd/pkg/infra/azure_resource_types.go [hunk 4]
```diff
@@ -13,6 +13,7 @@ const (
 	AzureResourceTypeStorageAccount          AzureResourceType = "Microsoft.Storage/storageAccounts"
 	AzureResourceTypeKeyVault                AzureResourceType = "Microsoft.KeyVault/vaults"
 	AzureResourceTypeAppConfig               AzureResourceType = "Microsoft.AppConfiguration/configurationStores"
+	AzureResourceTypeAPIM                    AzureResourceType = "Microsoft.ApiManagement/service"
```
Comment (vhvb1989) [inline cli/azd/pkg/infra/azure_resource_types.go:16]: This should be already added in main.  (Added it to display APIM as created resource).
You might want to rebase and look for `AzureResourceTypeApim` at line 26

File: cli/azd/pkg/infra/provisioning/bicep/bicep_provider.go [hunk 5]
```diff
@@ -282,6 +282,13 @@ func (p *BicepProvider) Destroy(
 				return
 			}
 
+			asyncContext.SetProgress(&DestroyProgress{Message: "Getting API Management Services to purge", Timestamp: time.Now()})
+			aPIManagements, err := p.getAPIManagementsToPurge(ctx, groupedResources)
```
Comment (vhvb1989) [inline cli/azd/pkg/infra/provisioning/bicep/bicep_provider.go:286]:
```suggestion
			apiManagements, err := p.getAPIManagementsToPurge(ctx, groupedResources)
```

File: cli/azd/pkg/infra/provisioning/bicep/bicep_provider.go [hunk 6]
```diff
@@ -585,6 +599,30 @@ func (p *BicepProvider) getAppConfigsToPurge(
 	return configs, nil
 }
 
+func (p *BicepProvider) getAPIManagementsToPurge(
```
Comment (vhvb1989) [inline cli/azd/pkg/infra/provisioning/bicep/bicep_provider.go:602]:
```suggestion
func (p *BicepProvider) getApiManagementsToPurge(
```

File: cli/azd/pkg/infra/provisioning/bicep/bicep_provider.go [hunk 7]
```diff
@@ -585,6 +599,30 @@ func (p *BicepProvider) getAppConfigsToPurge(
...
+	ctx context.Context,
+	groupedResources map[string][]azcli.AzCliResource,
+) ([]*azcli.AzCliAPIM, error) {
```
Comment (vhvb1989) [inline cli/azd/pkg/infra/provisioning/bicep/bicep_provider.go:605]:
```suggestion
) ([]*azcli.AzCliApim, error) {
```
It is my understanding that we don't use all-caps for such cases.

File: cli/azd/pkg/tools/azcli/apim.go [hunk 8]
```diff
@@ -0,0 +1,85 @@
+package azcli
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/Azure/azure-sdk-for-go/sdk/resourcemanager/apimanagement/armapimanagement"
+)
+
+type AzCliAPIM struct {
```
Comment (vhvb1989) [inline cli/azd/pkg/tools/azcli/apim.go:10]: Check if we should call it `AzCliApim` instead.  
Just to keep it the same way as we name other tools' structs

**Pull request merged**